addon/model.js

```diff
@@ -167,21 +163,45 @@ function resolveValue(key, value, modelName, store, schema, model, parentIdx) {
 
 // ie an array of nested models
 function resolvePlainArray(key, value, modelName, store, schema, model) {
-  if (value == null) {
-    return new Array(0);
+  let resolvedArray = new Array(0);
+  if (value && value.length > 0) {
+    resolvedArray = value.map((value, idx) =>
+      resolveValue(key, value, modelName, store, schema, model, idx)
+    );
   }
 
-  return value.map((value, idx) => resolveValue(key, value, modelName, store, schema, model, idx));
+  return M3PlainArray.create({
+    content: A(resolvedArray),
+    _replace(idx, removeAmt, newItems) {
+      // resolve new items
+      newItems = newItems.map(item => resolveValue(key, item, modelName, store, schema, model));
+
+      // Update content
+      this.content.replace(idx, removeAmt, newItems);
+
+      // Set attribute in model data and
+      // update model state
+      // and changedAttributes object
+      model._setAttribute(key, this.content);
+    },
+  });
 }
 
-function resolveRecordArray(store, references) {
+function resolveRecordArray(store, model, key, references) {
   let recordArrayManager = store._recordArrayManager;
 
   let array = M3RecordArray.create({
     modelName: '-ember-m3',
     content: A(),
     store: store,
     manager: recordArrayManager,
+    _replace(idx, removeAmt, newModels) {
```

Comment:
as with the array for nested models, it'll be clearer if we pass in context rather than a `_replace` function;

something like

```js
M3RecordArray.create({
  /* existing fields ... */
 model,
});
```